### PR TITLE
When all NS IPs are throttled, there is a problem with the recursor's cachehit calculation.

### PR DESCRIPTION
**Pull request by zjs604381586**
### Short description

When all NS IPs are throttled, there is a problem with the cachehit calculation of recursorcache. When res is RCode::ServFail, the hit statistics of the recursor cache should also be cachemiss, not cachehit.
### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

File: pdns/recursordist/pdns_recursor.cc [hunk 1]
```diff
@@ -1846,7 +1846,7 @@ void startDoResolve(void* arg) // NOLINT(readability-function-cognitive-complexi
     }
 
     if (comboWriter->d_mdp.d_header.opcode == static_cast<unsigned>(Opcode::Query)) {
-      if (resolver.d_outqueries != 0 || resolver.d_authzonequeries != 0) {
+      if (res == RCode::ServFail || resolver.d_outqueries != 0 || resolver.d_authzonequeries != 0) {
```
Comment (rgacogne) [inline pdns/recursordist/pdns_recursor.cc:1849]: While I agree with that the current behaviour is not right, we cannot simply rely on the rcode. We do serve `ServFail` answers from the cache, for example when a DNSSEC proof is broken, so a `ServFail` is not always a cache miss.

Comment (omoerbeek) [reply pdns/recursordist/pdns_recursor.cc:1849]: Indeed, will look into a better condition. 

Also a nte to @zjs604381586: we prefer PRs that do not contain merge commits Start with a clean, up-to-date master, branch and only add commits relevant to the fix. Then use the branch to submit the PR.

Comment (zjs604381586) [reply pdns/recursordist/pdns_recursor.cc:1849]: > Indeed, will look into a better condition.
> 
> Also a nte to @zjs604381586: we prefer PRs that do not contain merge commits Start with a clean, up-to-date master, branch and only add commits relevant to the fix. Then use the branch to submit the PR.

I see, thank you

**Comment by omoerbeek** [comment]
Superseded by #13497 which has (I think) the proper fix and no spurious commits.

**Comment by zjs604381586** [comment]
> Superseded by #13497 which has (I think) the proper fix and no spurious commits.

Great, this modification looks good

**Comment by zjs604381586** [comment]
@omoerbeek There is another question. If we only determine the number of **d_throttledqueries**, when the **query blocked by 'dont-query'** is triggered in **SyncRes::throttledOrBlocked()** function, do we need to perform the operation of **d_throttledqueries++**?

**Pull request closed without merging**